runtime/major_gc.c

```diff
@@ -687,21 +686,24 @@ static void clean_slice (intnat work)
 
 static void sweep_slice (intnat work)
 {
-  char *hp;
+  char *hp, *sweep_hp, *limit;
   header_t hd;
 
   caml_gc_message (0x40, "Sweeping %"
                    ARCH_INTNAT_PRINTF_FORMAT "d words\n", work);
+  sweep_hp = caml_gc_sweep_hp;
+  limit = sweep_chunk + Chunk_size(sweep_chunk);
   while (work > 0){
-    if (caml_gc_sweep_hp < sweep_limit){
-      hp = caml_gc_sweep_hp;
+    if (sweep_hp < limit){
+      caml_prefetch(sweep_hp + 4096);
```

Comment:
IIUC this will prefetch the cache line at `sweep_hp` + 4k. Isn't this likely to conflict with the cache line that contains `sweep_hp` itself? I know caches are associative, but the associativity is rather low, so what is the probability of evicting `sweep_hp` itself when we still need it? Would it be hard to benchmark some variation of this number (for example 4032)?

Reply:
Excellent point, I'll try that.